### ISSUE TITLE
Potential creation of invalid reference

### ISSUE THREAD

**Issue by 197g** [ababo/dtb#2]
The reader my construct a reference of `&Header` even when the underlying blob is too small. This may lead to UB and the compiler would also not miscompile if it were to remove the check for header length that comes afterwards, even though it does not seem to exploit that fact at the moment.

https://github.com/ababo/dtb/blob/3ca29eff4e29df50a45713bf954edcfb493ab5f7/src/reader.rs#L309

**Comment by ababo** [comment]
@HeroicKatora Can you share some concrete example causing this?

**Comment by 197g** [comment]
Any dtb with correct magic longer than 4 bytes but shorter than 40. There is no actual symptom though, as llvm doesn't exploit that fact (yet). It's a theoretical flaw that should be fixed but isn't critical.

**Comment by ababo** [comment]
@HeroicKatora Feel free to make a PR if you wish :-)

**Comment by ababo** [comment]
@HeroicKatora I don't see a problem here. Try to apply the following diff:
```diff
diff --git a/src/reader.rs b/src/reader.rs
index b662271..78de722 100644
--- a/src/reader.rs
+++ b/src/reader.rs
@@ -476,6 +476,7 @@ mod tests {
     }
 
     test_read_dtb!(test_bad_magic, BadMagic);
+    test_read_dtb!(test_small_header, UnexpectedEndOfBlob);
     test_read_dtb!(test_unexpected_end_of_blob, UnexpectedEndOfBlob);
     test_read_dtb!(test_bad_version, BadVersion);
     test_read_dtb!(test_unsupported_comp_version, UnsupportedCompVersion);
diff --git a/src/test_dtb/small_header.dtb b/src/test_dtb/small_header.dtb
new file mode 100644
index 0000000..3b08e79
--- /dev/null
+++ b/src/test_dtb/small_header.dtb
@@ -0,0 +1 @@
+<D0>^M<FE><ED>
\ No newline at end of file
```
You'll see it passes.

**Comment by 197g** [comment]
> There is no actual symptom though, as llvm doesn't exploit that fact (yet).

The usual way to point out that a theoretical flaw is UB according to Rust's memory model would be to run the tests through [miri](https://github.com/rust-lang/miri/) however both the file io and the assumption that `Vec<u8>` aligns its data suitable for `Reader::read` are not supported by the interpreter.

**Comment by 197g** [comment]
Also, on further review I found #4 which is much more critical. Please address it first.

**Comment by ababo** [comment]
So you mean that having a reference without the actual dereferencing of the potentially missing fields are dangerous? Can you share a reference to a memory model doc stating that?

**Comment by 197g** [comment]
Yes. It is dangerous because reference, unlike raw pointers, insert an assertion to llvm that allows it to add spurious reads to anywhere within the pointee. Since there is no *definitive, final* definition of Rust's current memory model,  [see this issue](https://github.com/rust-lang/unsafe-code-guidelines/issues/77) which also goes further into references and their semantics.

-----

On another note, it is a non sequitur to require a document that states this is dangerous. On the contrary, it is or should be the obligation of the progammer using `unsafe` to provide reference to documents proving it is *legal*- The instinct to reach for `unsafe` when something is not possible in the same way as in C is not good practice for Rust.

**Comment by ababo** [comment]
Fixed in #3
